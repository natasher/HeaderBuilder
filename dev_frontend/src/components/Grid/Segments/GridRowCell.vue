<template>
  <draggable
    :class      = "classObject"
    :style      = "`width: calc( ${ cellWidth }% - 4px )`"
    :data-title = "dataTitle"
    :key        = "rowKey"
    v-model     = "itemsModel"
    @change     = "onCellChange"
    :options    = "{
      group: {
        name: 'items',
      },
      disabled: ( getGridStatus === 'off' || getGridStatus === 'auto' )
                  ? true
                  : false,
      scroll: true,
      scrollSensitivity: 70,
      scrollSpeed: 10,
      animation: 150,
    }"
    data-placeholder = "Drop item here" >

    <base-item
      v-for   = "( item, index ) in itemsModel"
      :row    = "row"
      :cellPos= "cellPos"
      :class  = "{ 'disabled': ( getGridStatus === 'off' ) }"
      :item   = "item"
      :key    = "item.uuid"
      :cellId = "index" />

  </draggable>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'
import { capitalize } from '../../../filters/index'
import BaseItem  from '../../BaseItem.vue'
import draggable from 'vuedraggable'

export default {

  name: 'GridRowCell',

  props: {
    /**
     * Which row current component instance represents:
     *    [ 'actionBar', 'firstRow', 'secondRow' ]
     */
    row: {
      type    : String,
      required: true,
    },
    /**
     * Which current celll instance represents:
     *    [ 'all', 'left', 'center', 'right' ]
     */
    cellPos: {
      type    : String,
      required: true,
    },
  },

  components: {
    BaseItem,
    draggable,
  },

  data() {
    return {
      /**
       * @model
       */
      itemsModelName: '',
      rowKey        : '',
      dataTitle     : '',
      tmpItems      : void 0,
    }
  },

  computed: {
    ...mapGetters( 'ui', [
      'getGridStatus',
      'getLayoutPosition',
    ]),

    classObject: function() {
      return `hb-col hbc-${ this.cellPos }`
    },

    itemsModel: {
      get() {
        return this.$store.getters[ `items/get${ this.itemsModelName }` ]
      },
      set(items) {
        this.tmpItems = items
      },
    },

    cellWidth: function() {
      if (
        this.getLayoutPosition === 'left' || this.getLayoutPosition === 'right'
      ) {
        return '100'
      } else {
        return '33.3'
      }
    },
  },

  methods: {
    ...mapActions( 'cell', [ 'onChange' ]),

    onCellChange(event) {
      this.onChange({
        event    : event,
        row      : this.row,
        cellPos  : this.cellPos,
        tmpItems : this.tmpItems,
      })
    },
  },

  /** Lifecycle methods: */
  created() {
    const row  = this.row
    const cell = this.cellPos

    /**
     * @model
     */
    this.itemsModelName = `${ capitalize( row ) }Items${ capitalize( cell ) }`
    this.rowKey         = `${ row }${ capitalize( cell ) }`
    this.dataTitle      = `align ${ cell }`
  },
}
</script>
