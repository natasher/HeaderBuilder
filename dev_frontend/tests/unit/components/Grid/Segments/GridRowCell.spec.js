import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex          from 'vuex'
import merge         from 'lodash.merge'
import draggable     from 'vuedraggable'
import GridRowCell   from '../../../../../src/components/Grid/Segments/GridRowCell.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

const testItems = [
                {
                  "name": "social",
                  "uuid": "toWXiE1dIS",
                },
                {
                  "name": "menuIcon",
                  "uuid": "alBWLJBYfR"
                },
                {
                  "name": "button",
                  "uuid": "AJDV7xJMAN"
                }
              ]

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      ui: {
        namespaced: true,
        getters: {
          getGridStatus: jest.fn(() => { return 'custom' }),
          getLayoutPosition: jest.fn(() => { return 'top' }),
        }
      },
      cell: {
        namespaced: true,
        actions: {
          onChange: jest.fn()
        }
      },
      items: {
        namespaced: true,
        getters: {
          getActionBarItemsCenter: jest.fn(() => { return [] })
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    propsData: {
      row    : 'actionBar',
      cellPos: 'center'
    },
    stubs: {
      draggable,
    }
  }

  return shallowMount( GridRowCell, merge( defaultMountingOptions, overrides ))
}

describe('GridRowCell.vue', () => {

  describe('snapshots:', () => {

    test('should match without items', () => {
      const wrapper = createWrapper()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match with three items', () => {
      const getActionBarItemsCenter = jest.fn(() => { return testItems })
      const store = createStore({
        modules: {
          items: {
            getters: {
              getActionBarItemsCenter,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match with three items and grid status is `off`', () => {
      const getActionBarItemsCenter = jest.fn(() => { return testItems })
      const store = createStore({
        modules: {
          ui: {
            getters: {
              getGridStatus: jest.fn(() => { return 'off' }),
            },
          },
          items: {
            getters: {
              getActionBarItemsCenter,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

  describe('Dynamically created attributes (testable)', () => {

    test('should properly render draggable classes', () => {
      const cellPos = 'all'
      const wrapper = createWrapper({
        propsData: {
          cellPos,
        }
      })

      expect( wrapper.classes() ).toContain(`hbc-${ cellPos }`)
    })

    test('should properly set data-title on component', () => {
      const cellPos = 'all'
      const wrapper = createWrapper({
        propsData: {
          cellPos,
        }
      })

      expect( wrapper.attributes()[ 'data-title' ] ).toBe( `align ${ cellPos }` )
    })

  })

  describe('Rendered items', () => {

    test('should render proper count of items', () => {
      const getActionBarItemsCenter = jest.fn(() => { return testItems })
      const store = createStore({
        modules: {
          items: {
            getters: {
              getActionBarItemsCenter,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      const items = wrapper.findAll('baseitem-stub')

      expect( items.length ).toBe( testItems.length )
    })

    test('should pass row attribute to base items', () => {
      expect.assertions( 3 )

      const row = 'actionBar'
      const getActionBarItemsCenter = jest.fn(() => { return testItems })
      const store = createStore({
        modules: {
          items: {
            getters: {
              getActionBarItemsCenter,
            }
          }
        }
      })
      const wrapper = createWrapper({
        store,
        propsData: {
          row,
        }
      })
      const items = wrapper.findAll('baseitem-stub')

      for (let i = 0; i < testItems.length; i++) {
        expect( items.at( i ).attributes().row ).toBe( row )
      }
    })

    test('should pass cell position attribute to base items', () => {
      expect.assertions( 3 )

      const cellPos = 'center'
      const getActionBarItemsCenter = jest.fn(() => { return testItems })
      const store = createStore({
        modules: {
          items: {
            getters: {
              getActionBarItemsCenter,
            }
          }
        }
      })
      const wrapper = createWrapper({
        store,
        propsData: {
          cellPos,
        }
      })
      const items = wrapper.findAll('baseitem-stub')

      for (let i = 0; i < testItems.length; i++) {
        expect( items.at( i ).attributes().cellpos ).toBe( cellPos )
      }
    })

    test('should pass item object to base items', () => {
      expect.assertions( 3 )

      const getActionBarItemsCenter = jest.fn(() => { return testItems })
      const store = createStore({
        modules: {
          items: {
            getters: {
              getActionBarItemsCenter,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      const items = wrapper.findAll('baseitem-stub')

      for (let i = 0; i < testItems.length; i++) {
        expect( items.at( i ).attributes().item ).toBe( "[object Object]" )
      }
    })

    test('should pass item index in cell', () => {
      expect.assertions( 3 )

      const getActionBarItemsCenter = jest.fn(() => { return testItems })
      const store = createStore({
        modules: {
          items: {
            getters: {
              getActionBarItemsCenter,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      const items = wrapper.findAll('baseitem-stub')

      for (let i = 0; i < testItems.length; i++) {
        expect( items.at( i ).attributes().cellid ).toBe( `${ i }` )
      }
    })

    test('should put `disabled` class when grid status is `off`', () => {
      const getActionBarItemsCenter = jest.fn(() => { return testItems })

      const store = createStore({
        modules: {
          ui: {
            getters: {
              getGridStatus: jest.fn(() => { return 'off' }),
            },
          },
          items: {
            getters: {
              getActionBarItemsCenter,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      const items = wrapper.findAll('baseitem-stub')

      for (let i = 0; i < testItems.length; i++) {
        expect( items.at( i ).classes() ).toContain( 'disabled' )
      }
    })

  })

})