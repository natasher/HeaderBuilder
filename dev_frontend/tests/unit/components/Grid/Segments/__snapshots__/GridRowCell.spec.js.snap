// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridRowCell.vue snapshots: should match with three items 1`] = `
<div
  class="hb-col hbc-center"
  data-placeholder="Drop item here"
  data-title="align center"
>
  <base-item-stub
    cellid="0"
    cellpos="center"
    class=""
    item="[object Object]"
    row="actionBar"
  />
  <base-item-stub
    cellid="1"
    cellpos="center"
    class=""
    item="[object Object]"
    row="actionBar"
  />
  <base-item-stub
    cellid="2"
    cellpos="center"
    class=""
    item="[object Object]"
    row="actionBar"
  />
</div>
`;

exports[`GridRowCell.vue snapshots: should match with three items and grid status is \`off\` 1`] = `
<div
  class="hb-col hbc-center"
  data-placeholder="Drop item here"
  data-title="align center"
>
  <base-item-stub
    cellid="0"
    cellpos="center"
    class="disabled"
    item="[object Object]"
    row="actionBar"
  />
  <base-item-stub
    cellid="1"
    cellpos="center"
    class="disabled"
    item="[object Object]"
    row="actionBar"
  />
  <base-item-stub
    cellid="2"
    cellpos="center"
    class="disabled"
    item="[object Object]"
    row="actionBar"
  />
</div>
`;

exports[`GridRowCell.vue snapshots: should match without items 1`] = `
<div
  class="hb-col hbc-center"
  data-placeholder="Drop item here"
  data-title="align center"
/>
`;
